frontend/src/App.js

```diff
@@ -57,26 +57,64 @@ export class App extends React.Component {
 
   handleChange = (change) => {
     switch (change.operation) {
-      case "addElement":
-        this.addElement(change.data);
+      case "addComponent":
+        this.addComponent(change.data);
         break;
       case "moveComponent":
         this.moveComponent(change.data);
         break;
       case "deleteComponent":
         this.deleteComponent(change.data);
         break;
+
+      case "addConnection":
+        this.addConnection(change.data);
+        break;
+      case "updateConnection":
+        this.updateConnection(change.data);
+        break;
+      case "deleteConnection":
+        this.deleteConnection(change.data);
+        break;
       default:
         console.error("unhandled change:", change.data);
         break;
     }
   };
 
-  addElement = (elementDetails) => {
+  addComponent = (componentDetails) => {
+    this.setState({
+      diagramData: {
+        ...this.state.diagramData,
+        components: [...this.state.diagramData.components, componentDetails],
+      },
+    });
+  };
+
+  addConnection = (connectionDetails) => {
+    console.log("connectionDetails:", connectionDetails);
+    this.setState({
+      diagramData: {
+        ...this.state.diagramData,
+        connections: [...this.state.diagramData.connections, connectionDetails],
+      },
+    });
+  };
+
+  updateConnection = (connectionDetails) => {
+    const { connections } = this.state.diagramData;
+    const targetIndex = connections.findIndex(
+      (element) => element.id === connectionDetails.id
+    );
+
     this.setState({
       diagramData: {
         ...this.state.diagramData,
-        components: [...this.state.diagramData.components, elementDetails],
+        connections: [
```

Comment:
wouldn't it be simpler to just do this in two lines, i.e.
`connections[targetIndex] = connectionDetails`
then in the `this.setState` you say
`connections: connections`?

Reply:
you can't directly mutate `connections`. Every change made to `connections` must be done via `setState`

Reply:
Fair enough. I guess when you do the destructuring, you are assigning by reference and not by value, so if you edit `const connections` then you are editing `this.state.diagramData.connections`.

Reply:
yes, that is correct. All non-scalar types in javascript are passed by reference (and all of them are essentially objects)